Ignore cleanup failures when applying ASG updates

mariash opened the pull request:
After new ASG chain is created and we fail to clean up old ASG chain we
log the error but continue on and consider our new ASG chain as
successful and save it in the in-memory map for subsequent comparison.
This is to avoid re-applying the same chain. As well as saving it in a
map as a desired chain to avoid deleting it right after we create it.

Original ASG chain fails to be deleted because cascading deletion of
referenced chains might fail. In the observed case we saw that
referenced chain is the netout-...-log chain that is getting referenced
by the new ASG chain now.

Signed-off-by: Marc Paquette <mpaquette@vmware.com>

mariash committed:
Ignore cleanup failures when applying ASG updates

After new ASG chain is created and we fail to clean up old ASG chain we
log the error but continue on and consider our new ASG chain as
successful and save it in the in-memory map for subsequent comparison.
This is to avoid re-applying the same chain. As well as saving it in a
map as a desired chain to avoid deleting it right after we create it.

Original ASG chain fails to be deleted because cascading deletion of
referenced chains might fail. In the observed case we saw that
referenced chain is the netout-...-log chain that is getting referenced
by the new ASG chain now.

Signed-off-by: Marc Paquette <mpaquette@vmware.com>

geofffranks commented on src/code.cloudfoundry.org/vxlan-policy-agent/converger/converger_test.go:
```diff
@@ -603,48 +603,118 @@ var _ = Describe("Single Poll Cycle", func() {
-					}
-					i++
-					return fmt.Sprintf("%s-with-new-suffix", e.Chain.Prefix), err
-				}
```
this looks like it was moved below, but the comment is now abandoned. should that get moved or cleaned up as well?

mariash replied:
cool, thanks for the notice

geofffranks reviewed (changes requested):
Looks good aside from the orphaned comment

MarcPaquette commented on src/code.cloudfoundry.org/vxlan-policy-agent/converger/converger.go:
```diff
@@ -155,12 +155,14 @@ func (m *SinglePollCycle) SyncASGsForContainers(containers ...string) error {
 				if err != nil {
-					errors = multierror.Append(errors, fmt.Errorf("enforce-asg: %s", err))
+					if _, ok := err.(*enforcer.CleanupErr); ok {
+						m.logger.Error("failed-to-cleanup", err)
```
Why don't we use `multierror.Append(errors, fmt.Errorf("failed-to-cleanup %s", err))` here?

MarcPaquette replied:
Why don't we use a `errors = multierror.Append(errors, fmt.Errorf("failed-to-cleanup: %s", err))` here?

Commit:
Keep reject rule in the parent netout chain

When ASG chain is created and jump rule to it is added in the parent
netout chain we clean up the parent netout chain. We want to keep the
REJECT rule in the parent chain so that no traffic is allowed after the
ASG chain in case of any erros while creating ASG chain.

Signed-off-by: Maria Shaldybin <mariash@vmware.com>

mariash committed:
Return the cleanup error when enforcing asg rules

Update local in-memory set with new asg chain if cleanup fails but still
return it as an error. The caller will log it anyway.

Signed-off-by: Marc Paquette <mpaquette@vmware.com>

MarcPaquette reviewed (approved):
LGTM
